bson/src/main/org/bson/AbstractBsonWriter.java

```diff
@@ -752,36 +753,58 @@ public void close() {
     @Override
     public void pipe(final BsonReader reader) {
         notNull("reader", reader);
-        reader.readStartDocument();
-        writeStartDocument();
-        while (reader.readBsonType() != BsonType.END_OF_DOCUMENT) {
-            writeName(reader.readName());
-            pipeValue(reader);
+        pipeDocument(reader, null);
+    }
+
+    /**
+     * Reads a single document from the given BsonReader and writes it to this, appending the given extra elements to the document.
+     *
+     * @param reader the source of the document
+     * @param extraElements the extra elements to append to the document
+     * @since 3.6
+     */
+    public void pipe(final BsonReader reader, final List<BsonElement> extraElements) {
```

Comment:
Unfortunate introduction of a new public method in a patch release.  I don't see a way around it.  Unlike the original commit I didn't introduce this pipe method into the BsonWriter interface.  That would be a bridge too far...